generic/piranha/src/models/rule.rs

```diff
@@ -179,50 +185,309 @@ impl Rule {
-    self.query = query;
+  pub(crate) fn update_query(&mut self, substitutions: &HashMap<String, String>) {
+    self.query = substitute_tags(self.query(), substitutions);
   }
 
   pub(crate) fn name(&self) -> String {
     String::from(&self.name)
   }
-}
 
-#[derive(Deserialize, Debug, Clone, Hash, PartialEq, Eq)]
-pub struct Constraint {
-  /// Scope in which the constraint query has to be applied
-  matcher: String,
-  /// The Tree-sitter queries that need to be applied in the `matcher` scope
-  queries: Vec<String>,
-}
-
-impl Constraint {
-  pub(crate) fn queries(&self) -> &[String] {
-    &self.queries
+  // Apply all the `rules` to the node, parent, grand parent and great grand parent.
+  // Short-circuit on the first match.
+  pub(crate) fn get_edit_for_context(
+    source_code_unit: &SourceCodeUnit, previous_edit_start: usize, previous_edit_end: usize,
+    rules_store: &mut RuleStore, rules: &Vec<Rule>,
+  ) -> Option<Edit> {
+    // Context contains -  the changed node in the previous edit, its's parent, grand parent and great grand parent
+    let context = || {
+      get_context(
+        source_code_unit.root_node(),
+        previous_edit_start,
+        previous_edit_end,
+      )
+    };
+    for rule in rules {
+      for ancestor in &context() {
+        if let Some(edit) = rule.get_edit(&source_code_unit.clone(), rules_store, *ancestor, false)
+        {
+          return Some(edit);
+        }
+      }
+    }
+    None
   }
 
-  pub(crate) fn matcher(&self) -> String {
-    String::from(&self.matcher)
+  /// Gets the first match for the rule in `self`
+  pub(crate) fn get_edit(
+    &self, source_code_unit: &SourceCodeUnit, rule_store: &mut RuleStore, node: Node,
+    recursive: bool,
+  ) -> Option<Edit> {
+    // Get all matches for the query in the given scope `node`.
+    let all_query_matches = node.get_all_matches_for_query(
+      source_code_unit.code(),
+      rule_store.get_query(&self.get_query()),
+      recursive,
+      Some(self.replace_node()),
+    );
+
+    // Return the first match that satisfies constraint of the rule
+    for (range, tag_substitutions) in all_query_matches {
+      let matched_node = get_node_for_range(
+        source_code_unit.root_node(),
+        range.start_byte,
+        range.end_byte,
+      );
+      if matched_node.satisfies_constraint(
+        source_code_unit.clone(),
+        self,
+        &tag_substitutions,
+        rule_store,
+      ) {
+        let replacement = substitute_tags(self.replace(), &tag_substitutions);
+        return Some(Edit::new(
+          range,
+          replacement,
+          self.clone(),
+          tag_substitutions,
+        ));
+      }
+    }
+    None
   }
 }
 
-pub trait RuleHelper {
-  /// replaces the all the occurrences of keys (of `substitutions` map) in the string with its corresponding value.
-  fn substitute_tags(&self, substitutions: &HashMap<String, String>) -> String;
-}
+mod test {
```

Comment:
do these tests need to reside within the same file?

Reply:
I fixed this using this http://xion.io/post/code/rust-unit-test-placement.html